module/spl/spl-kmem.c

```diff
@@ -133,6 +133,93 @@ strfree(char *str)
 }
 EXPORT_SYMBOL(strfree);
 
+/* kernel compat for <4.13 */
+#ifndef __GFP_RETRY_MAYFAIL
+#define	__GFP_RETRY_MAYFAIL	__GFP_REPEAT
+#endif
+
+void *
+spl_kvmalloc(size_t size, gfp_t lflags, const char *caller)
+{
+
+	void *ptr;
+	
+#ifdef HAVE_KVMALLOC
+	/*
+	 * GFP_KERNEL allocations can safely use kvmalloc which may
+	 * improve performance by avoiding a) high latency caused by
+	 * vmalloc's on-access allocation, b) performance loss due to
+	 * MMU memory address mapping and c) vmalloc locking overhead.
+	 * This has the side-effect that the slab statistics will
+	 * incorrectly report this as a vmem allocation, but that is
+	 * purely cosmetic.
+	 */
+	if ((lflags & GFP_KERNEL) == GFP_KERNEL) {
+		ptr = (kvmalloc(size, lflags));
+		if (!ptr)
+			printk(KERN_WARNING "spl_kvmalloc: kvmalloc called by %s with GFP_KERNEL=%d, GFP_ATOMIC=%d, failed\n",
+			    caller, ((lflags & GFP_KERNEL) == GFP_KERNEL) ? 1 : 0,
+					((lflags & GFP_ATOMIC) == GFP_ATOMIC) ? 1 : 0);
+	}
```

Comment:
I do not understand this. If `kvmalloc()` returns a valid pointer, why are we not returning that and instead falling through to the generic code for implementing the same thing? It looks like a double allocation and memory leak to me.

Also, if we are going to implement the same thing for kernels that lack `kvmalloc()`, why bother calling the kernel symbol at all? We could have a uniform behavior across all kernels by not bothering with `kvmalloc()`?

Reply:
you can ignore all that printk stuff.... this was just a desparate attempt of debugging....

Reply:
Oh, well... ouch. I see, what you meant... removing that return was *not* intended.